src/util/antlr/ANTLRErrorHandling.cpp

```diff
@@ -2,38 +2,76 @@
+
+#include "util/Exception.h"
+
+// _____________________________________________________________________________
+ExceptionMetadata generateMetadata(antlr4::Parser* parser,
+                                   antlr4::Token* offendingToken, size_t line,
+                                   size_t charPositionInLine) {
+  return {
+      parser->getTokenStream()->getTokenSource()->getInputStream()->toString(),
+      offendingToken->getStartIndex(), offendingToken->getStopIndex(), line,
+      charPositionInLine};
+}
+
+// _____________________________________________________________________________
+ExceptionMetadata generateMetadata(antlr4::Lexer* lexer, size_t line,
+                                   size_t charPositionInLine) {
+  return {lexer->getInputStream()->toString(), lexer->tokenStartCharIndex,
+          lexer->tokenStartCharIndex, line, charPositionInLine};
+}
 
 // _____________________________________________________________________________
 ExceptionMetadata generateMetadata(antlr4::Recognizer* recognizer,
-                                   antlr4::Token* offendingToken,
-                                   [[maybe_unused]] size_t line,
-                                   [[maybe_unused]] size_t charPositionInLine) {
-  return {((antlr4::CommonTokenStream*)recognizer->getInputStream())
-              ->getTokenSource()
-              ->getInputStream()
-              ->toString(),
-          offendingToken->getStartIndex(), offendingToken->getStopIndex()};
+                                   antlr4::Token* offendingToken, size_t line,
+                                   size_t charPositionInLine) {
+  // The `antlr4::Recognizer` interface does not provide enough information to
+  // obtain a copy of the input string. Do this switching here to obtain that.
+  // As of ANTLRv4 `antlr4::Parser` and `antlr4::Lexer` are the only subclasses
+  // of `antlr4::Recognizer`.
+  if (auto* parser = dynamic_cast<antlr4::Parser*>(recognizer)) {
+    AD_CHECK(offendingToken != nullptr);
+    return generateMetadata(parser, offendingToken, line, charPositionInLine);
+  } else if (auto* lexer = dynamic_cast<antlr4::Lexer*>(recognizer)) {
+    // If the recognizer is a Lexer this means that the error was a Lexer error.
+    // In that case the tokens are not yet available and `offendingToken` is
+    // NULL.
+    return generateMetadata(lexer, line, charPositionInLine);
+  } else {
+    AD_FAIL();  // Should be unreachable.
+  }
 }
 
 // _____________________________________________________________________________
 ExceptionMetadata generateMetadata(antlr4::ParserRuleContext* ctx) {
-  return {ctx->getStart()->getInputStream()->toString(),
-          ctx->getStart()->getStartIndex(), ctx->getStop()->getStopIndex()};
+  auto start = ctx->getStart();
+  return {start->getInputStream()->toString(), ctx->getStart()->getStartIndex(),
+          ctx->getStop()->getStopIndex(), start->getLine(),
+          start->getCharPositionInLine()};
```

Comment:
Nice, so this interface already exists without manually calculating anything?

Reply:
Yes. My fear previously was that I would have to calculate in the other direction (line+position in line -> startIndex). But this also thankfully turned out to not be needed.